pico/configdriven/services/src/main/java/io/helidon/pico/configdriven/services/AbstractConfiguredServiceProvider.java

```diff
@@ -0,0 +1,832 @@
+                .addQualifier(DefaultQualifierAndValue.createNamed(instanceId))
+                .build();
+
+        // override our service info
+        instance.serviceInfo(newServiceInfo);
+        instance.picoServices(Optional.of(picoServices()));
+        instance.rootProvider(this);
+
+        if (logger().isLoggable(System.Logger.Level.DEBUG)) {
+            logger().log(System.Logger.Level.DEBUG, "config instance successfully initialized: "
+                    + id() + ":" + newServiceInfo.qualifiers());
+        }
+
+        return Optional.of(instance);
+    }
+
+    /**
+     * Return true if this service is driven to activation during startup (and provided it has some config).
+     * See {@link io.helidon.pico.configdriven.ConfiguredBy#drivesActivation()} and
+     * see {@link io.helidon.builder.config.ConfigBean#drivesActivation()} for more.
+     * @return true if this service is driven to activation during startup
+     */
+    // note: overridden by the service if disabled at the ConfiguredBy service level
+    protected boolean drivesActivation() {
+        return metaConfigBeanInfo().drivesActivation();
+    }
+
+    /**
+     * Configurable services by their very nature are not compile-time bindable during application creation.
+     *
+     * @return empty, signaling that we are not bindable
+     */
+    @Override
+    public Optional<ServiceProviderBindable<T>> serviceProviderBindable() {
+        return Optional.empty();
+    }
+
+    /**
+     * The special comparator for ordering config bean instance ids.
+     *
+     * @return the special comparator for ordering config bean instance ids
+     */
+    static Comparator<String> configBeanComparator() {
+        return BEAN_INSTANCE_ID_COMPARATOR;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <C extends Config, CB> Optional<CB> toConfigBean(C config,
+                                                            Class<CB> configBeanType) {
+        CB configBean = (CB) toConfigBean(config);
+        assert (configBeanType.isInstance(configBean)) : configBean;
+        return Optional.of(configBean);
+    }
+
+    /**
+     * See {@link #configBeanComparator()}.
+     */
+    static class CBInstanceComparator implements Comparator<String>, Serializable {
+        @Override
+        public int compare(String str1,
+                           String str2) {
+            if (DefaultConfigBeanRegistry.DEFAULT_INSTANCE_ID.equals(str1)) {
+                return -1 * Integer.MAX_VALUE;
+            } else if (DefaultConfigBeanRegistry.DEFAULT_INSTANCE_ID.equals(str2)) {
+                return Integer.MAX_VALUE;
+            }
+            return str1.compareTo(str2);
+        }
+    }
+
+    static InternalConfigBeanRegistry resolveConfigBeanRegistry() {
+        BasicConfigBeanRegistry cbr = ConfigBeanRegistryHolder.configBeanRegistry().orElse(null);
+        if (cbr == null) {
```

Comment:
this should use an early return, the current method is confusing.

pseudo:
if (null) {
    return null;
}
if (!cbr instanceof I....) {
 throw
}
return cbr;

Reply:
pushed